In [73]:
import pandas as pd
import numpy as np
import datetime as dt

from eda_utilities import print_eda_stats

### Read the Dockless Mobility data and run basic null data analysis. 
##### Dockless mobility data is located on the s3 bucket at https://s3.amazonaws.com/sameera-bucket-1/dockless_mobility/raw_data/Austin_Dockless_Vehicle_Trips.csv
##### Weather data is located on the s3 bucket at https://s3.amazonaws.com/sameera-bucket-1/dockless_mobility/raw_data/Austin_Bergstom_Airport_Weather.csv

In [2]:
df = pd.read_csv("https://s3.amazonaws.com/sameera-bucket-1/dockless_mobility/raw_data/Austin_Dockless_Vehicle_Trips.csv")

/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (13) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [5]:
dockless_vehicles_columns = df.columns

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2746505 entries, 0 to 2746504
Data columns (total 20 columns):
ID                          object
Device ID                   object
Vehicle Type                object
Trip Duration               float64
Trip Distance               float64
Start Time                  object
End Time                    object
Modified Date               object
Month                       float64
Hour                        float64
Day of Week                 float64
Council District (Start)    float64
Council District (End)      float64
Origin Cell ID              object
Destination Cell ID         object
Year                        float64
Start Latitude              float64
Start Longitude             float64
End Latitude                float64
End Longitude               float64
dtypes: float64(12), object(8)
memory usage: 419.1+ MB


In [7]:
print_eda_stats(df, dockless_vehicles_columns, "Austin Dockless Mobility Data")

+--------------------------+---------+-------------+---------------+--------------+-------------------+
|       Column Name        |  dtype  | Total_Count | Unique_Values | df_totalrows | Null Value counts |
+--------------------------+---------+-------------+---------------+--------------+-------------------+
|            ID            |  object |   2746505   |    2746505    |   2746505    |         0         |
|        Device ID         |  object |   2746505   |     26999     |   2746505    |         0         |
|       Vehicle Type       |  object |   2746504   |       3       |   2746505    |         1         |
|      Trip Duration       | float64 |   2746504   |     10521     |   2746505    |         1         |
|      Trip Distance       | float64 |   2746504   |     16125     |   2746505    |         1         |
|        Start Time        |  object |   2746504   |     25694     |   2746505    |         1         |
|         End Time         |  object |   2746504   |     25768  

In [9]:
#WITH FOUR DECIMAL PLACES (DEFAULT)
df.describe().transpose().to_csv('summary_stats.csv', sep=',')

np.round(df.describe(), 2).T

count      mean          std          min      25%      50%      75%           max
Trip Duration             2746504.0    671.94      7441.55    -86161.00   221.00   388.00   753.00  1.149160e+07
Trip Distance             2746504.0  58032.67  10421635.18 -16724106.00   343.00   887.00  1738.00  2.147484e+09
Month                     2746504.0      8.23         3.50         1.00     7.00     9.00    11.00  1.200000e+01
Hour                      2746504.0     14.44         4.38         0.00    12.00    15.00    18.00  2.300000e+01
Day of Week               2746504.0      3.17         2.06         0.00     1.00     3.00     5.00  6.000000e+00
Council District (Start)  2746474.0      7.73         2.65         0.00     9.00     9.00     9.00  1.000000e+01
Council District (End)    2746474.0      7.68         2.67         0.00     9.00     9.00     9.00  1.000000e+01
Year                      2746504.0   2018.16         0.37      2018.00  2018.00  2018.00  2018.00  2.019000e+03
Start Latitude            2713821.0     30.19         1.60         0.00    30.26    30.27    30.28  8.229000e+01
Start Longitude           2713821.0    -97.45         5.45      -122.68   -97.75   -97.74   -97.74  1.141300e+02
End Latitude              2724693.0     30.19         1.61       -43.52    30.26    30.27    30.28  5.224000e+01
End Longitude             2724693.0    -97.45         5.50      -122.68   -97.75   -97.74   -97.74  1.750300e+02

In [8]:
weather_df = pd.read_csv("https://s3.amazonaws.com/sameera-bucket-1/dockless_mobility/raw_data/Austin_Bergstom_Airport_Weather.csv")

In [15]:
weather_columns = weather_df.columns
weather_columns

Index(['STATION', 'NAME', 'LATITUDE', 'LONGITUDE', 'ELEVATION', 'DATE', 'AWND',
       'AWND_ATTRIBUTES', 'PGTM', 'PGTM_ATTRIBUTES', 'PRCP', 'PRCP_ATTRIBUTES',
       'SNOW', 'SNOW_ATTRIBUTES', 'SNWD', 'SNWD_ATTRIBUTES', 'TAVG',
       'TAVG_ATTRIBUTES', 'TMAX', 'TMAX_ATTRIBUTES', 'TMIN', 'TMIN_ATTRIBUTES',
       'WDF2', 'WDF2_ATTRIBUTES', 'WDF5', 'WDF5_ATTRIBUTES', 'WSF2',
       'WSF2_ATTRIBUTES', 'WSF5', 'WSF5_ATTRIBUTES', 'WT01', 'WT01_ATTRIBUTES',
       'WT02', 'WT02_ATTRIBUTES', 'WT03', 'WT03_ATTRIBUTES', 'WT04',
       'WT04_ATTRIBUTES', 'WT05', 'WT05_ATTRIBUTES', 'WT06', 'WT06_ATTRIBUTES',
       'WT08', 'WT08_ATTRIBUTES'],
      dtype='object')

In [14]:
print_eda_stats(weather_df, weather_columns, "Austin Weather Data")

+-----------------+---------+-------------+---------------+--------------+-------------------+
|   Column Name   |  dtype  | Total_Count | Unique_Values | df_totalrows | Null Value counts |
+-----------------+---------+-------------+---------------+--------------+-------------------+
|     STATION     |  object |     436     |       1       |     436      |         0         |
|       NAME      |  object |     436     |       1       |     436      |         0         |
|     LATITUDE    | float64 |     436     |       1       |     436      |         0         |
|    LONGITUDE    | float64 |     436     |       1       |     436      |         0         |
|    ELEVATION    | float64 |     436     |       1       |     436      |         0         |
|       DATE      |  object |     436     |      436      |     436      |         0         |
|       AWND      | float64 |     435     |       73      |     436      |         1         |
| AWND_ATTRIBUTES |  object |     435     |       

In [16]:
weather_columns_to_drop = ['STATION', 'NAME', 'LATITUDE', 'LONGITUDE', 'ELEVATION',
       'AWND_ATTRIBUTES', 'PGTM', 'PGTM_ATTRIBUTES', 'PRCP_ATTRIBUTES',
       'SNOW_ATTRIBUTES', 'SNWD', 'SNWD_ATTRIBUTES',
       'TAVG_ATTRIBUTES', 'TMAX_ATTRIBUTES', 'TMIN_ATTRIBUTES',
       'WDF2', 'WDF2_ATTRIBUTES', 'WDF5', 'WDF5_ATTRIBUTES', 'WSF2',
       'WSF2_ATTRIBUTES', 'WSF5', 'WSF5_ATTRIBUTES', 'WT01', 'WT01_ATTRIBUTES',
       'WT02', 'WT02_ATTRIBUTES', 'WT03', 'WT03_ATTRIBUTES', 'WT04',
       'WT04_ATTRIBUTES', 'WT05', 'WT05_ATTRIBUTES', 'WT06', 'WT06_ATTRIBUTES',
       'WT08', 'WT08_ATTRIBUTES']


weather_df.drop(weather_columns_to_drop, axis=1, inplace=True)

In [18]:
weather_df[weather_df['DATE'] == '2019-03-12']

DATE  AWND  PRCP  SNOW  TAVG  TMAX  TMIN
435  2019-03-12   NaN   NaN   NaN  59.0   NaN   NaN

In [29]:
weather_columns = weather_df.columns
# Drop the '2019-03-12' because it was not available 
# weather_df.drop(weather_columns_to_drop, axis=1, inplace=True)
print_eda_stats(weather_df, weather_columns)

+-------------+---------+-------------+---------------+--------------+-------------------+
| Column Name |  dtype  | Total_Count | Unique_Values | df_totalrows | Null Value counts |
+-------------+---------+-------------+---------------+--------------+-------------------+
|     DATE    |  object |     436     |      436      |     436      |         0         |
|     AWND    | float64 |     436     |       72      |     436      |         0         |
|     PRCP    | float64 |     436     |       52      |     436      |         0         |
|     SNOW    | float64 |     436     |       1       |     436      |         0         |
|     TAVG    | float64 |     436     |       62      |     436      |         0         |
|     TMAX    | float64 |     436     |       70      |     436      |         0         |
|     TMIN    | float64 |     436     |       59      |     436      |         0         |
+-------------+---------+-------------+---------------+--------------+-------------------+

#### Initial Analysis of the weather data revealed some null values of the data. The data was not available for the day the data was downloaded and also there was an additional missing values. Both '2019-03-01' and '2019-03-02' had missing TAVG data. From the current analysis I concluded the data was missing at random (MAR) so decided to impede the null values with the data from the previous record.

In [28]:

weather_df[weather_df.isnull().any(axis=1)]


Empty DataFrame
Columns: [DATE, AWND, PRCP, SNOW, TAVG, TMAX, TMIN]
Index: []

In [27]:
weather_df.fillna(method='ffill', inplace=True)

In [26]:
weather_df[weather_df['DATE'] == '2019-03-11']

DATE   AWND  PRCP  SNOW  TAVG  TMAX  TMIN
434  2019-03-11  11.86   0.0   0.0  63.0  67.0  58.0

In [30]:
df.columns

Index(['ID', 'Device ID', 'Vehicle Type', 'Trip Duration', 'Trip Distance',
       'Start Time', 'End Time', 'Modified Date', 'Month', 'Hour',
       'Day of Week', 'Council District (Start)', 'Council District (End)',
       'Origin Cell ID', 'Destination Cell ID', 'Year', 'Start Latitude',
       'Start Longitude', 'End Latitude', 'End Longitude'],
      dtype='object')

In [31]:
df['Vehicle Type'].unique()

array(['scooter', 'bicycle', nan], dtype=object)

In [34]:
dfnull = df[df.isnull().any(axis=1)]

In [33]:
# mobility_columns_to_drop = ['ID', 'Device ID', 'Vehicle Type', 'Trip Duration', 'Trip Distance',
#        'Start Time', 'End Time', 'Modified Date', 'Month', 'Hour',
#        'Day of Week', 'Council District (Start)', 'Council District (End)',
#        'Origin Cell ID', 'Destination Cell ID', 'Year', 'Start Latitude',
#        'Start Longitude', 'End Latitude', 'End Longitude']

df[df['Vehicle Type'].isnull()]

ID  \
825895  6865731e-5e4b-4819-9435-4e62e09a652d   

                                   Device ID Vehicle Type  Trip Duration  \
825895  08c8c852-257c-49ad-a14f-727bfc79626d          NaN            NaN   

        Trip Distance Start Time End Time           Modified Date  Month  \
825895            NaN        NaN      NaN  12/13/2018 03:36:12 PM    NaN   

        Hour  Day of Week  Council District (Start)  Council District (End)  \
825895   NaN          NaN                       0.0                     0.0   

       Origin Cell ID Destination Cell ID  Year  Start Latitude  \
825895            NaN                 NaN   NaN             NaN   

        Start Longitude  End Latitude  End Longitude  
825895              NaN           NaN            NaN

In [35]:
dfnull['Vehicle Type'].unique()

array(['scooter', 'bicycle', nan], dtype=object)

In [36]:
dfnull

ID  \
145      0003ef39-d6c6-42be-b4bd-717d5e213f80   
267      00080fda-ef07-4c72-81bb-b331ac087a31   
336      000a8aa1-21c7-49f7-93ad-fc8889a05314   
406      000cc84e-90ff-4315-bf45-da33035f28ac   
414      000d0e79-2763-449c-80cc-9058686ab90e   
453      000e6942-3a96-4384-93af-97ec721c4cc0   
461      000ea052-ae29-4e6b-891f-bccdc0b8c481   
588      00126ad5-74ad-4073-98f3-95bc158de3a4   
616      001313bb-d8a3-4ba7-ba7b-5c79cd2a00d1   
630      001377ba-2c67-4384-8f0b-edfaaeb8e65a   
661      0014d55d-1e40-4b7a-a0d3-61a66167554d   
671      00155b70-3466-42ee-b046-cb306b308a36   
677      0015a86e-446c-4be0-9529-ff9a683cc859   
720      0016dab4-c296-47bf-9863-aa5ac28ea8b9   
746      0017b1f5-a65e-44b3-8635-288280b20705   
767      001861f1-bc98-4c87-ac4a-a6a48bdea753   
797      00192b32-5626-455c-92e3-a99a7c97398f   
894      001c5303-4826-4696-bbb4-7399973abf75   
974      001e8570-19c2-4b1f-8e92-f9bb4a6049a7   
1020     001fc942-b390-4ec4-92d5-a49c2763ce65   
1024     001fdb97-19ea-40e7-8295-1059d165babc   
1061     00212342-c576-40f7-9863-b33afe8ef9cf   
1063     00212a04-17da-4eb6-8559-e3aec5ff4cde   
1072     00218c29-47e8-4e80-9fcd-c2c1c60a1540   
1174     00255020-61c1-4dff-bff6-d09272336739   
1224     0026ec13-1b8a-4f46-a9cd-1722a22c3289   
1294     0028cf23-9615-4d90-b91c-424142b1e643   
1322     0029c334-fd8c-492e-a48f-b7a81d0f1e6d   
1371     002b14dc-82a4-4821-8d33-1218120676d4   
1380     002b4b66-7c01-484a-918e-1976cf94d677   
...                                       ...   
2577656  cace5569-3aed-43a5-a965-d4bdd233f8bc   
2577657  39aeb439-baad-427a-9e7d-9e6e2d9dae07   
2577658  134e04e2-7ba7-47d5-a42f-75e9a55f56b0   
2577659  f00a6c08-31b3-477b-8a85-bf0d3ca9447e   
2577660  518bdd4e-3385-41cc-a91c-82f2cb6f42a6   
2577661  492e603f-dc62-42e3-a483-e16564743796   
2577662  66f53043-4d1c-43fd-9458-e5459cea3551   
2577663  1a36520b-950a-494d-bd6a-54c8aa43bd11   
2577664  b9bdea38-51ff-4f22-9327-a8317174a695   
2577665  52770b6e-4e3c-4c01-8222-99244fce87f6   
2577666  87f3ce42-1ba8-464d-ad6c-ed693c218df7   
2577667  8e33cc0d-259f-489f-85c5-08448c30ba5f   
2577668  1423a41e-a719-4d54-9deb-02b77fd1f097   
2577669  549c9f71-1da7-4bf4-b8c8-55e376197438   
2577670  d9b2843d-51c5-4809-a383-c7480d933af8   
2577671  f0095e32-e091-475f-8f44-6483e0884af4   
2577672  1642f1f4-53ff-4f5c-8a11-7d8727ae959e   
2577679  3ef57177-a93e-4016-97f3-3136829878a2   
2577680  a79234dc-1f99-4bee-9d2b-d040ed847c0f   
2577681  338cef81-4b3e-40c8-8d9d-93e2938c7455   
2577682  fd9e25bc-4c62-4bbc-8e53-e01402404d59   
2577683  7ea27686-ea7c-4f00-8cf4-de8bbd691949   
2577684  eac6b2cd-ae54-4e87-a460-ea3c4f84b62c   
2577685  112e1625-5728-4709-b0b4-27b8d021a37e   
2577686  86398f1a-e851-4ce2-94e7-aeb4d4181923   
2577687  71e122e3-38ed-40db-b63b-d46753844f3e   
2577688  073c5dda-77da-4acb-86dc-97f6964b162f   
2577689  ce7afb26-683b-4d24-ada0-3313a48333ae   
2577690  39a882c2-ab52-402d-8722-94fbddcb08cb   
2577691  d260b657-7f23-4840-9c78-dccc3719cd8e   

                                    Device ID Vehicle Type  Trip Duration  \
145      f4cf290d-8ee2-444a-a98e-cc2e8ae0f971      scooter          418.0   
267      ad1e67ea-ae3b-4f55-b158-32d61c9198cd      scooter           63.0   
336      09b79c18-52e6-4801-9df7-1936bf5fb25e      scooter          441.0   
406      c97c4f15-95a8-40a2-8ac9-814aa4600923      scooter          208.0   
414      658f7e34-9e47-4b7e-9bb9-c7239859122f      scooter          414.0   
453      8ea6275e-7122-48f2-89c8-b6d93f553204      scooter          408.0   
461      c362be77-8235-400d-a33a-a4cf2c742be6      scooter          313.0   
588      b6e3766e-2166-4ce6-808e-182846f4523f      scooter          582.0   
616      add78e92-55ff-421c-a0f0-b2fe66e41f6b      scooter          556.0   
630      1d816fcf-d573-48cd-ae68-e6cf364e2432      scooter         5287.0   
661      3d5aee50-b38d-4076-a96a-1431bbe148d7      scooter         1094.0   
671      63723ec1-49a0-4396-90e5-0ba5808cb561      scooter          643.0   
677

In [37]:
dfnull.to_csv("Mobility_Data_Null_Values.csv", index=False)

In [42]:
origin_cell_null = dfnull[(dfnull['Origin Cell ID'] == 'OUT_OF_BOUNDS') 
                          | (dfnull['Destination Cell ID'] == 'OUT_OF_BOUNDS')]






In [43]:
origin_cell_null.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 42430 entries, 145 to 2408238
Data columns (total 20 columns):
ID                          42430 non-null object
Device ID                   42430 non-null object
Vehicle Type                42430 non-null object
Trip Duration               42430 non-null float64
Trip Distance               42430 non-null float64
Start Time                  42430 non-null object
End Time                    42430 non-null object
Modified Date               42430 non-null object
Month                       42430 non-null float64
Hour                        42430 non-null float64
Day of Week                 42430 non-null float64
Council District (Start)    42430 non-null float64
Council District (End)      42430 non-null float64
Origin Cell ID              42430 non-null object
Destination Cell ID         42430 non-null object
Year                        42430 non-null float64
Start Latitude              13058 non-null float64
Start Longitude             

In [44]:
dfnull.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46695 entries, 145 to 2577691
Data columns (total 20 columns):
ID                          46695 non-null object
Device ID                   46695 non-null object
Vehicle Type                46694 non-null object
Trip Duration               46694 non-null float64
Trip Distance               46694 non-null float64
Start Time                  46694 non-null object
End Time                    46694 non-null object
Modified Date               46695 non-null object
Month                       46694 non-null float64
Hour                        46694 non-null float64
Day of Week                 46694 non-null float64
Council District (Start)    46664 non-null float64
Council District (End)      46664 non-null float64
Origin Cell ID              44607 non-null object
Destination Cell ID         44607 non-null object
Year                        46694 non-null float64
Start Latitude              14011 non-null float64
Start Longitude             

In [45]:
dfnull.shape

(46695, 20)

In [46]:
origin_cell_null.shape

(42430, 20)

In [47]:
df.shape

(2746505, 20)

#### Initial Analysis of the Scooter Data revealed that are 46695 rows are null out of 2746505 rows. Out of them 2088 of them have either Origin Cell ID or Destination Cell ID has 'OUT_OF_BOUNDS' value. I will be follow up with City of Austin transportation department regarding null values. For the first pass I will be deleting all the rows which have 'OUT_OF_BOUNDS' in the Origin or Destination Cell ID as it is required for the data analysis.

In [49]:
df.drop(df.index[(df['Origin Cell ID'] == 'OUT_OF_BOUNDS') 
                          | (df['Destination Cell ID'] == 'OUT_OF_BOUNDS')], inplace = True)

In [50]:
df.shape

(2695253, 20)

#### Since data of interest currently is only scooter data I have decided all the data that is not of type scooter. 'Vehicle Type' != 'scooter'

In [52]:
df[df['Vehicle Type'] != 'scooter']


df.drop(df.index[(df['Vehicle Type'] != 'scooter')], inplace = True)

In [53]:
df.shape

(2619407, 20)

#### Data size changed from 2746505 - > 2695253 (After removing the Cell ID's = 'OUT_OF_BOUNDS') -> 2619407 (After removing the Vehicle Type != 'scooter'

In [54]:
print_eda_stats(df, dockless_vehicles_columns)

+--------------------------+---------+-------------+---------------+--------------+-------------------+
|       Column Name        |  dtype  | Total_Count | Unique_Values | df_totalrows | Null Value counts |
+--------------------------+---------+-------------+---------------+--------------+-------------------+
|            ID            |  object |   2619407   |    2619407    |   2619407    |         0         |
|        Device ID         |  object |   2619407   |     26124     |   2619407    |         0         |
|       Vehicle Type       |  object |   2619407   |       1       |   2619407    |         0         |
|      Trip Duration       | float64 |   2619407   |      9799     |   2619407    |         0         |
|      Trip Distance       | float64 |   2619407   |     15464     |   2619407    |         0         |
|        Start Time        |  object |   2619407   |     25513     |   2619407    |         0         |
|         End Time         |  object |   2619407   |     25599  

In [56]:
df[(df['Origin Cell ID'].isnull()) | (df['Destination Cell ID'].isnull())]

ID  \
746      0017b1f5-a65e-44b3-8635-288280b20705   
1072     00218c29-47e8-4e80-9fcd-c2c1c60a1540   
1599     003362d8-d42a-42fd-930d-526260013d7e   
2357     004c5f0b-9217-4071-a9f8-f4d5349456c0   
2976     006111e9-bec3-42de-9a58-6543083d4bee   
4285     008c697b-e6a8-45bb-8592-0215f8def5d2   
5320     00ac9773-1e7e-4073-a018-975123f23a4f   
5596     00b5b66f-7c60-46a4-ac0b-ec7dc6f098a7   
6014     00c3bdb5-0463-4d71-aa68-bdab86a675db   
6833     00df0264-a5b6-4e06-b2c3-3b93f3123b52   
9313     0130172f-6cb5-49ad-953d-448e5a4c29f5   
10419    0153c663-b9ae-4f31-aa5e-601c7f14bf4b   
11152    016cf6c0-7ae2-4b7d-965a-d2ecb3d5f128   
11302    01714fae-67f9-490b-9762-de29061ad9cf   
11316    08766fe9-7ea9-4c4b-bf03-0ec7c17add55   
11615    017c459f-358f-4cf3-9666-befded7a3b0f   
11635    017cecb9-55f6-428f-ab60-05bdfee98559   
13029    01aa77e8-3b56-4630-86d3-affd59bcc3b1   
13241    01b092de-2f75-467e-bb09-789ba4ab89ec   
13539    01ba9d3c-e638-427c-9b21-9c687db02161   
13808    01c36f0e-2d7c-49c6-b915-1301fcea0050   
13828    0ad7cb41-fb85-47f4-be08-99fbdd6f42ce   
14055    01ca8d8a-cf33-44b8-bacd-0a882b9a931e   
20070    028df2e4-3399-4c35-bf5a-599fe76423d5   
20858    02a6cd68-406a-4aab-b2bd-d5f33f690383   
21413    02b97cdf-06eb-4f35-928c-ffb2244eb263   
22957    02ed16bc-cea4-4c21-a494-c4ad8261efa1   
24071    03110f9a-c9f0-43d1-b9eb-98ac82d7e610   
24519    03207be6-085f-4414-bcc6-62b4dcd13cc7   
25404    033e7162-12c4-4576-a80e-72de3112abd4   
...                                       ...   
2577656  cace5569-3aed-43a5-a965-d4bdd233f8bc   
2577657  39aeb439-baad-427a-9e7d-9e6e2d9dae07   
2577658  134e04e2-7ba7-47d5-a42f-75e9a55f56b0   
2577659  f00a6c08-31b3-477b-8a85-bf0d3ca9447e   
2577660  518bdd4e-3385-41cc-a91c-82f2cb6f42a6   
2577661  492e603f-dc62-42e3-a483-e16564743796   
2577662  66f53043-4d1c-43fd-9458-e5459cea3551   
2577663  1a36520b-950a-494d-bd6a-54c8aa43bd11   
2577664  b9bdea38-51ff-4f22-9327-a8317174a695   
2577665  52770b6e-4e3c-4c01-8222-99244fce87f6   
2577666  87f3ce42-1ba8-464d-ad6c-ed693c218df7   
2577667  8e33cc0d-259f-489f-85c5-08448c30ba5f   
2577668  1423a41e-a719-4d54-9deb-02b77fd1f097   
2577669  549c9f71-1da7-4bf4-b8c8-55e376197438   
2577670  d9b2843d-51c5-4809-a383-c7480d933af8   
2577671  f0095e32-e091-475f-8f44-6483e0884af4   
2577672  1642f1f4-53ff-4f5c-8a11-7d8727ae959e   
2577679  3ef57177-a93e-4016-97f3-3136829878a2   
2577680  a79234dc-1f99-4bee-9d2b-d040ed847c0f   
2577681  338cef81-4b3e-40c8-8d9d-93e2938c7455   
2577682  fd9e25bc-4c62-4bbc-8e53-e01402404d59   
2577683  7ea27686-ea7c-4f00-8cf4-de8bbd691949   
2577684  eac6b2cd-ae54-4e87-a460-ea3c4f84b62c   
2577685  112e1625-5728-4709-b0b4-27b8d021a37e   
2577686  86398f1a-e851-4ce2-94e7-aeb4d4181923   
2577687  71e122e3-38ed-40db-b63b-d46753844f3e   
2577688  073c5dda-77da-4acb-86dc-97f6964b162f   
2577689  ce7afb26-683b-4d24-ada0-3313a48333ae   
2577690  39a882c2-ab52-402d-8722-94fbddcb08cb   
2577691  d260b657-7f23-4840-9c78-dccc3719cd8e   

                                    Device ID Vehicle Type  Trip Duration  \
746      d7f1ccf7-ac15-4c08-bf95-acfe10769f01      scooter          340.0   
1072     02e3ac3b-5db7-437c-a559-b8ff40889c83      scooter          161.0   
1599     d7f1ccf7-ac15-4c08-bf95-acfe10769f01      scooter          176.0   
2357     aaefed18-0043-4b52-ac8c-d64b6431f94a      scooter         2723.0   
2976     7d268ff8-01c3-4d35-8312-a700703f34d8      scooter          192.0   
4285     6097686d-9fed-46e1-a90b-1e8857c8485b      scooter          122.0   
5320     e185d474-7c4a-4c08-b9e1-d4cf53ca10e4      scooter          145.0   
5596     be130773-572d-4f42-8f0e-b120982dc709      scooter         2942.0   
6014     eb8328d4-6b95-4d49-982b-95fbc18ff0bf      scooter         1763.0   
6833     73c52998-a325-480a-b96a-5c2a3a41e02a      scooter          900.0   
9313     4400e894-a0d1-4e95-823d-1a7adcf667b8      scooter          651.0   
10419    3789522e-899f-4c8f-ac83-d8a298492f35      scooter           26.0   
111

In [57]:

df.drop(df.index[(df['Origin Cell ID'].isnull()) | (df['Destination Cell ID'].isnull())], inplace = True)

In [58]:
df.shape

(2617320, 20)

In [59]:
print_eda_stats(df, dockless_vehicles_columns)

+--------------------------+---------+-------------+---------------+--------------+-------------------+
|       Column Name        |  dtype  | Total_Count | Unique_Values | df_totalrows | Null Value counts |
+--------------------------+---------+-------------+---------------+--------------+-------------------+
|            ID            |  object |   2617320   |    2617320    |   2617320    |         0         |
|        Device ID         |  object |   2617320   |     26096     |   2617320    |         0         |
|       Vehicle Type       |  object |   2617320   |       1       |   2617320    |         0         |
|      Trip Duration       | float64 |   2617320   |      9794     |   2617320    |         0         |
|      Trip Distance       | float64 |   2617320   |     15455     |   2617320    |         0         |
|        Start Time        |  object |   2617320   |     25506     |   2617320    |         0         |
|         End Time         |  object |   2617320   |     25590  

#### Data size changed from 2746505 - > 2695253 (After removing the Cell ID's = 'OUT_OF_BOUNDS') -> 2619407 (After removing the Vehicle Type != 'scooter' - > 2617320 (After removing the Cell ID's null). After all the cleaning there are still 1329 rows with Start Latitude, Longitude to be null and 1158 rows with End Latitude and Longitude to be null

In [60]:
df.drop(df.index[(df['Start Latitude'].isnull()) | (df['End Latitude'].isnull())], inplace = True)

In [61]:
df.shape

(2615143, 20)

In [62]:
print_eda_stats(df, dockless_vehicles_columns)

+--------------------------+---------+-------------+---------------+--------------+-------------------+
|       Column Name        |  dtype  | Total_Count | Unique_Values | df_totalrows | Null Value counts |
+--------------------------+---------+-------------+---------------+--------------+-------------------+
|            ID            |  object |   2615143   |    2615143    |   2615143    |         0         |
|        Device ID         |  object |   2615143   |     26094     |   2615143    |         0         |
|       Vehicle Type       |  object |   2615143   |       1       |   2615143    |         0         |
|      Trip Duration       | float64 |   2615143   |      9773     |   2615143    |         0         |
|      Trip Distance       | float64 |   2615143   |     15441     |   2615143    |         0         |
|        Start Time        |  object |   2615143   |     25504     |   2615143    |         0         |
|         End Time         |  object |   2615143   |     25588  

In [63]:
df.columns

Index(['ID', 'Device ID', 'Vehicle Type', 'Trip Duration', 'Trip Distance',
       'Start Time', 'End Time', 'Modified Date', 'Month', 'Hour',
       'Day of Week', 'Council District (Start)', 'Council District (End)',
       'Origin Cell ID', 'Destination Cell ID', 'Year', 'Start Latitude',
       'Start Longitude', 'End Latitude', 'End Longitude'],
      dtype='object')

In [64]:
mobility_columns_to_drop = ['ID', 'Device ID', 'Vehicle Type', 'Modified Date']

In [65]:
df.drop(mobility_columns_to_drop, axis=1, inplace=True)

In [66]:
df.head(5)

Trip Duration  Trip Distance              Start Time  \
0          358.0          915.0  06/29/2018 05:30:00 PM   
1          226.0          839.0  11/30/2018 11:00:00 AM   
2          324.0         1206.0  10/02/2018 03:15:00 PM   
3         1096.0            0.0  09/02/2018 04:45:00 PM   
4          408.0         1144.0  09/28/2018 11:45:00 AM   

                 End Time  Month  Hour  Day of Week  Council District (Start)  \
0  06/29/2018 05:30:00 PM    6.0  17.0          5.0                       5.0   
1  11/30/2018 11:00:00 AM   11.0  11.0          5.0                       9.0   
2  10/02/2018 03:30:00 PM   10.0  15.0          2.0                       9.0   
3  09/02/2018 05:00:00 PM    9.0  16.0          0.0                       9.0   
4  09/28/2018 11:45:00 AM    9.0  11.0          5.0                       3.0   

   Council District (End) Origin Cell ID Destination Cell ID    Year  \
0                     9.0         013445              013916  2018.0   
1                     9.0         014556              015186  2018.0   
2                     9.0         015344              015032  2018.0   
3                     9.0         014707              014081  2018.0   
4                     3.0         016114              015642  2018.0   

   Start Latitude  Start Longitude  End Latitude  End Longitude  
0          30.263          -97.762        30.259        -97.754  
1          30.283          -97.745        30.284        -97.737  
2          30.285          -97.734        30.291        -97.738  
3          30.268          -97.743        30.276        -97.752  
4          30.242          -97.721        30.244        -97.729

In [67]:
weather_df.head(5)

DATE   AWND  PRCP  SNOW  TAVG  TMAX  TMIN
0  2018-01-01  16.78   0.0   0.0  28.0  32.0  24.0
1  2018-01-02  10.74   0.0   0.0  29.0  32.0  23.0
2  2018-01-03   3.13   0.0   0.0  31.0  56.0  15.0
3  2018-01-04   2.68   0.0   0.0  36.0  54.0  22.0
4  2018-01-05   4.47   0.0   0.0  49.0  61.0  35.0

In [69]:
weather_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 436 entries, 0 to 435
Data columns (total 7 columns):
DATE    436 non-null object
AWND    436 non-null float64
PRCP    436 non-null float64
SNOW    436 non-null float64
TAVG    436 non-null float64
TMAX    436 non-null float64
TMIN    436 non-null float64
dtypes: float64(6), object(1)
memory usage: 23.9+ KB


In [70]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2615143 entries, 0 to 2746504
Data columns (total 16 columns):
Trip Duration               float64
Trip Distance               float64
Start Time                  object
End Time                    object
Month                       float64
Hour                        float64
Day of Week                 float64
Council District (Start)    float64
Council District (End)      float64
Origin Cell ID              object
Destination Cell ID         object
Year                        float64
Start Latitude              float64
Start Longitude             float64
End Latitude                float64
End Longitude               float64
dtypes: float64(12), object(4)
memory usage: 339.2+ MB


In [71]:
weather_df.rename(columns={'AWND': 'AVG_DAILY_WIND_SPEED', 'PRCP': 'PRECIPITATION','TAVG': 'AVG_TEMPERATURE', 'TMAX': 'MAX_TEMPERATURE','TMIN': 'MIN_TEMPERATURE' }, inplace=True)

In [75]:
weather_df.info()

weather_df_1 = weather_df
weather_df_1

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 436 entries, 0 to 435
Data columns (total 7 columns):
DATE                    436 non-null object
AVG_DAILY_WIND_SPEED    436 non-null float64
PRECIPITATION           436 non-null float64
SNOW                    436 non-null float64
AVG_TEMPERATURE         436 non-null float64
MAX_TEMPERATURE         436 non-null float64
MIN_TEMPERATURE         436 non-null float64
dtypes: float64(6), object(1)
memory usage: 23.9+ KB


DATE  AVG_DAILY_WIND_SPEED  PRECIPITATION  SNOW  AVG_TEMPERATURE  \
0    2018-01-01                 16.78           0.00   0.0             28.0   
1    2018-01-02                 10.74           0.00   0.0             29.0   
2    2018-01-03                  3.13           0.00   0.0             31.0   
3    2018-01-04                  2.68           0.00   0.0             36.0   
4    2018-01-05                  4.47           0.00   0.0             49.0   
5    2018-01-06                  4.47           0.00   0.0             45.0   
6    2018-01-07                 10.29           0.00   0.0             59.0   
7    2018-01-08                 11.18           0.00   0.0             60.0   
8    2018-01-09                  2.91           0.00   0.0             48.0   
9    2018-01-10                  9.40           0.00   0.0             53.0   
10   2018-01-11                 12.53           0.00   0.0             61.0   
11   2018-01-12                 12.53           0.00   0.0             45.0   
12   2018-01-13                  8.05           0.00   0.0             38.0   
13   2018-01-14                  2.91           0.00   0.0             38.0   
14   2018-01-15                  8.72           0.10   0.0             45.0   
15   2018-01-16                 19.46           0.08   0.0             36.0   
16   2018-01-17                  7.61           0.00   0.0             26.0   
17   2018-01-18                  4.03           0.00   0.0             32.0   
18   2018-01-19                  2.91           0.01   0.0             44.0   
19   2018-01-20                  7.38           0.00   0.0             54.0   
20   2018-01-21                 10.51           0.00   0.0             63.0   
21   2018-01-22                  8.50           0.00   0.0             59.0   
22   2018-01-23                  3.58           0.00   0.0             46.0   
23   2018-01-24                  2.01           0.00   0.0             43.0   
24   2018-01-25                  6.04           0.00   0.0             42.0   
25   2018-01-26                  6.04           0.00   0.0             56.0   
26   2018-01-27                  8.05           0.06   0.0             64.0   
27   2018-01-28                  9.17           0.00   0.0             60.0   
28   2018-01-29                  7.61           0.00   0.0             50.0   
29   2018-01-30                  5.37           0.00   0.0             44.0   
..          ...                   ...            ...   ...              ...   
406  2019-02-11                  4.70           0.05   0.0             53.0   
407  2019-02-12                  8.95           0.00   0.0             59.0   
408  2019-02-13                  7.16           0.00   0.0             46.0   
409  2019-02-14                 12.30           0.00   0.0             60.0   
410  2019-02-15                  7.16           0.00   0.0             65.0   
411  2019-02-16                 10.51           0.00   0.0             55.0   
412  2019-02-17                 13.20           0.00   0.0             55.0   
413  2019-02-18                 15.43           0.00   0.0             52.0   
414  2019-02-19                 12.97           0.15   0.0             45.0   
415  2019-02-20                  6.49           0.00   0.0             48.0   
416  2019-02-21                  7.61           0.00   0.0             51.0   
417  2019-02-22                  6.71           0.03   0.0             53.0   
418  2019-02-23                  8.50           0.00   0.0             62.0   
419  2019-02-24                  9.40           0.00   0.0             55.0   
420  2019-02-25                  7.16           0.00   0.0             51.0   
421  2019-02-26                  5.59           0.04   0.0             58.0   
422  2019-02-27                  4.25           0.00   0.0             64.0   
423  2019-02-28                 12.08           0.04   0.0             50.0   
424  2019-03-01                  4.92           0.00   0.0   

In [77]:
# Convert the date in weather dataframe to datetime object

weather_df_1['DATE'] = weather_df_1['DATE'].apply(lambda x: dt.datetime.strptime(x,'%Y-%m-%d'))

In [78]:
weather_df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 436 entries, 0 to 435
Data columns (total 7 columns):
DATE                    436 non-null datetime64[ns]
AVG_DAILY_WIND_SPEED    436 non-null float64
PRECIPITATION           436 non-null float64
SNOW                    436 non-null float64
AVG_TEMPERATURE         436 non-null float64
MAX_TEMPERATURE         436 non-null float64
MIN_TEMPERATURE         436 non-null float64
dtypes: datetime64[ns](1), float64(6)
memory usage: 23.9 KB


In [79]:
weather_df_1.head(5)

DATE  AVG_DAILY_WIND_SPEED  PRECIPITATION  SNOW  AVG_TEMPERATURE  \
0 2018-01-01                 16.78            0.0   0.0             28.0   
1 2018-01-02                 10.74            0.0   0.0             29.0   
2 2018-01-03                  3.13            0.0   0.0             31.0   
3 2018-01-04                  2.68            0.0   0.0             36.0   
4 2018-01-05                  4.47            0.0   0.0             49.0   

   MAX_TEMPERATURE  MIN_TEMPERATURE  
0             32.0             24.0  
1             32.0             23.0  
2             56.0             15.0  
3             54.0             22.0  
4             61.0             35.0

#### Weather Data Details. Precipitation - Units of Measure -  inches, Snow - Units of Measure - inches, Avg_Daily_Wind_Speed - Units of Measure - MPH, AVG_TEMPERATURE, MAX_TEMPERATURE, MIN_TEMPERATURE- Units of Measure - Degrees Fahrenheit

In [88]:
df_1 = df

In [84]:
df_1.head(5)

Trip Duration  Trip Distance              Start Time  \
0          358.0          915.0  06/29/2018 05:30:00 PM   
1          226.0          839.0  11/30/2018 11:00:00 AM   
2          324.0         1206.0  10/02/2018 03:15:00 PM   
3         1096.0            0.0  09/02/2018 04:45:00 PM   
4          408.0         1144.0  09/28/2018 11:45:00 AM   

                 End Time  Month  Hour  Day of Week  Council District (Start)  \
0  06/29/2018 05:30:00 PM    6.0  17.0          5.0                       5.0   
1  11/30/2018 11:00:00 AM   11.0  11.0          5.0                       9.0   
2  10/02/2018 03:30:00 PM   10.0  15.0          2.0                       9.0   
3  09/02/2018 05:00:00 PM    9.0  16.0          0.0                       9.0   
4  09/28/2018 11:45:00 AM    9.0  11.0          5.0                       3.0   

   Council District (End) Origin Cell ID Destination Cell ID    Year  \
0                     9.0         013445              013916  2018.0   
1                     9.0         014556              015186  2018.0   
2                     9.0         015344              015032  2018.0   
3                     9.0         014707              014081  2018.0   
4                     3.0         016114              015642  2018.0   

   Start Latitude  Start Longitude  End Latitude  End Longitude  
0          30.263          -97.762        30.259        -97.754  
1          30.283          -97.745        30.284        -97.737  
2          30.285          -97.734        30.291        -97.738  
3          30.268          -97.743        30.276        -97.752  
4          30.242          -97.721        30.244        -97.729

In [89]:
df_1['Start Time'] = df_1['Start Time'].apply(lambda x: dt.datetime.strptime(x,'%m/%d/%Y %I:%M:%S %p'))
df_1['End Time'] = df_1['End Time'].apply(lambda x: dt.datetime.strptime(x,'%m/%d/%Y %I:%M:%S %p'))

In [90]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2615143 entries, 0 to 2746504
Data columns (total 18 columns):
Trip Duration               float64
Trip Distance               float64
Start Time                  datetime64[ns]
End Time                    datetime64[ns]
Month                       float64
Hour                        float64
Day of Week                 float64
Council District (Start)    float64
Council District (End)      float64
Origin Cell ID              object
Destination Cell ID         object
Year                        float64
Start Latitude              float64
Start Longitude             float64
End Latitude                float64
End Longitude               float64
START_DATE                  datetime64[ns]
END_DATE                    datetime64[ns]
dtypes: datetime64[ns](4), float64(12), object(2)
memory usage: 379.1+ MB


In [91]:
# df['just_date'] = df['dates'].dt.date
df_1.head(5)

Trip Duration  Trip Distance          Start Time            End Time  \
0          358.0          915.0 2018-06-29 17:30:00 2018-06-29 17:30:00   
1          226.0          839.0 2018-11-30 11:00:00 2018-11-30 11:00:00   
2          324.0         1206.0 2018-10-02 15:15:00 2018-10-02 15:30:00   
3         1096.0            0.0 2018-09-02 16:45:00 2018-09-02 17:00:00   
4          408.0         1144.0 2018-09-28 11:45:00 2018-09-28 11:45:00   

   Month  Hour  Day of Week  Council District (Start)  Council District (End)  \
0    6.0  17.0          5.0                       5.0                     9.0   
1   11.0  11.0          5.0                       9.0                     9.0   
2   10.0  15.0          2.0                       9.0                     9.0   
3    9.0  16.0          0.0                       9.0                     9.0   
4    9.0  11.0          5.0                       3.0                     3.0   

  Origin Cell ID Destination Cell ID    Year  Start Latitude  Start Longitude  \
0         013445              013916  2018.0          30.263          -97.762   
1         014556              015186  2018.0          30.283          -97.745   
2         015344              015032  2018.0          30.285          -97.734   
3         014707              014081  2018.0          30.268          -97.743   
4         016114              015642  2018.0          30.242          -97.721   

   End Latitude  End Longitude          START_DATE            END_DATE  
0        30.259        -97.754 2018-06-29 17:30:00 2018-06-29 17:30:00  
1        30.284        -97.737 2018-11-30 11:00:00 2018-11-30 11:00:00  
2        30.291        -97.738 2018-10-02 15:15:00 2018-10-02 15:30:00  
3        30.276        -97.752 2018-09-02 16:45:00 2018-09-02 17:00:00  
4        30.244        -97.729 2018-09-28 11:45:00 2018-09-28 11:45:00

In [98]:
df_1['START_DATE'] = df_1['Start Time'].dt.date
df_1['END_DATE'] = df_1['End Time'].dt.date

In [99]:
df_1.head(5)

Trip Duration  Trip Distance          Start Time            End Time  \
0          358.0          915.0 2018-06-29 17:30:00 2018-06-29 17:30:00   
1          226.0          839.0 2018-11-30 11:00:00 2018-11-30 11:00:00   
2          324.0         1206.0 2018-10-02 15:15:00 2018-10-02 15:30:00   
3         1096.0            0.0 2018-09-02 16:45:00 2018-09-02 17:00:00   
4          408.0         1144.0 2018-09-28 11:45:00 2018-09-28 11:45:00   

   Month  Hour  Day of Week  Council District (Start)  Council District (End)  \
0    6.0  17.0          5.0                       5.0                     9.0   
1   11.0  11.0          5.0                       9.0                     9.0   
2   10.0  15.0          2.0                       9.0                     9.0   
3    9.0  16.0          0.0                       9.0                     9.0   
4    9.0  11.0          5.0                       3.0                     3.0   

  Origin Cell ID Destination Cell ID    Year  Start Latitude  Start Longitude  \
0         013445              013916  2018.0          30.263          -97.762   
1         014556              015186  2018.0          30.283          -97.745   
2         015344              015032  2018.0          30.285          -97.734   
3         014707              014081  2018.0          30.268          -97.743   
4         016114              015642  2018.0          30.242          -97.721   

   End Latitude  End Longitude  START_DATE    END_DATE  
0        30.259        -97.754  2018-06-29  2018-06-29  
1        30.284        -97.737  2018-11-30  2018-11-30  
2        30.291        -97.738  2018-10-02  2018-10-02  
3        30.276        -97.752  2018-09-02  2018-09-02  
4        30.244        -97.729  2018-09-28  2018-09-28

In [103]:
df_1.Month.unique()

array([ 6, 11, 10,  9,  7,  4, 12,  8,  5,  1,  2])

In [101]:
# Convert the month, hour, Day of Week, Council District(Start, End), Year to integers instead of floats
int_columns = ['Month', 'Hour', 'Day of Week', 'Council District (Start)', 'Council District (End)', 'Year']
df_1[int_columns] = df_1[int_columns].astype(int)

In [102]:
df_1.head(5)

Trip Duration  Trip Distance          Start Time            End Time  \
0          358.0          915.0 2018-06-29 17:30:00 2018-06-29 17:30:00   
1          226.0          839.0 2018-11-30 11:00:00 2018-11-30 11:00:00   
2          324.0         1206.0 2018-10-02 15:15:00 2018-10-02 15:30:00   
3         1096.0            0.0 2018-09-02 16:45:00 2018-09-02 17:00:00   
4          408.0         1144.0 2018-09-28 11:45:00 2018-09-28 11:45:00   

   Month  Hour  Day of Week  Council District (Start)  Council District (End)  \
0      6    17            5                         5                       9   
1     11    11            5                         9                       9   
2     10    15            2                         9                       9   
3      9    16            0                         9                       9   
4      9    11            5                         3                       3   

  Origin Cell ID Destination Cell ID  Year  Start Latitude  Start Longitude  \
0         013445              013916  2018          30.263          -97.762   
1         014556              015186  2018          30.283          -97.745   
2         015344              015032  2018          30.285          -97.734   
3         014707              014081  2018          30.268          -97.743   
4         016114              015642  2018          30.242          -97.721   

   End Latitude  End Longitude  START_DATE    END_DATE  
0        30.259        -97.754  2018-06-29  2018-06-29  
1        30.284        -97.737  2018-11-30  2018-11-30  
2        30.291        -97.738  2018-10-02  2018-10-02  
3        30.276        -97.752  2018-09-02  2018-09-02  
4        30.244        -97.729  2018-09-28  2018-09-28

####  Trip duration - UOM in seconds, Trip distance - UOM in meters, Month - where 1 = January, etc.,  Hour - The hour of the day during which trip occurred, in local time (US/Central),  Day of Week - where Sunday = 0, and so on, 